### PR TITLE
refactor(resources): migrate WidgetList to TypeScript 

### PR DESCRIPTION
**Pull request by Trim21**
again

It looks like the element is not removed correctly. So the `_sorted` is getting bigger and bigger. so `_sorted.sort()` will cause leg(in `layout()`).

### CODE REVIEW

**Comment by Trim21** [comment]
I have tested it in browser with profiling, looks fine.

File: resources/widget_list.ts [hunk 1]
```diff
@@ -1,17 +1,40 @@
...
+  private _xinc1 = 1;
+  private _xinc2 = 0;
+  private _yinc1 = 0;
+  private _yinc2 = 1;
+  private _rowcolsize = 5;
+  private _maxnumber = 1000;
+  private _connected = false;
+  private rootElement: HTMLElement;
+  // bypass shadowRoot maybe null in base class
+  private shadow: ShadowRoot;
```
Comment (quisquous) [inline resources/widget_list.ts:24]: Would you mind explaining more why you can't use `this.shadowRoot`? What would it look like if you did?

**Review by quisquous — COMMENTED** [review]
```diff
diff --git a/resources/widget_list.ts b/resources/widget_list.ts
index ff58c943..e88a61f0 100644
--- a/resources/widget_list.ts
+++ b/resources/widget_list.ts
@@ -3,9 +3,7 @@
 // just switch over to using CSS grid.
 type Sorter = () => number;
 
-const getRandomInt = function(max: number) {
-  return Math.floor(Math.random() * Math.floor(max));
-};
+const getRandomInt = (max: number) => Math.floor(Math.random() * Math.floor(max));
 
 export default class WidgetList extends HTMLElement {
   private _nextId = 1;
@@ -22,6 +20,8 @@ export default class WidgetList extends HTMLElement {
   private _maxnumber = 1000;
   private _connected = false;
   private rootElement: HTMLElement;
+  // bypass shadowRoot maybe null in base class
+  private shadow: ShadowRoot;
 
   static get observedAttributes(): string[] {
     return ['toward', 'elementwidth', 'elementheight', 'rowcolsize', 'maxnumber'];
@@ -106,11 +106,8 @@ export default class WidgetList extends HTMLElement {
     root.innerHTML = `
       <div id="root" style="position: relative"></div>
     `;
-
-    if (this.shadowRoot === null)
-      throw new Error('shadowRoot of widget-list is null');
-
-    this.rootElement = this.shadowRoot.getElementById('root') as HTMLElement;
+    this.shadow = root;
+    this.rootElement = root.getElementById('root') as HTMLElement;
   }
 
   connectedCallback(): void {
@@ -219,14 +216,11 @@ export default class WidgetList extends HTMLElement {
     if (old)
       this.removeElement(name);
 
-    let sortKeyFn;
-    if (typeof sortKey === 'number') {
-      sortKeyFn = function() {
-        return sortKey;
-      };
-    } else {
+    let sortKeyFn: Sorter;
+    if (typeof sortKey === 'number')
+      sortKeyFn = () => sortKey;
+    else
       sortKeyFn = sortKey;
-    }
 
     this._nameToId[name] = id;
     this._elements[id] = sortKeyFn;
@@ -251,13 +245,23 @@ export default class WidgetList extends HTMLElement {
     const id = this._nameToId[name];
     if (!id)
       return;
-    const container = this.shadowRoot?.getElementById(`child${id}`);
-    if (container !== null && container !== undefined) {
-      const element = container.childNodes[0];
+    const container = this.shadow.getElementById(`child${id}`);
+    let element = undefined;
+    if (container) {
+      element = container.childNodes[0];
       this.rootElement.removeChild(container);
-      return element;
     }
-    return;
+
+    delete this._nameToId[name];
+    delete this._elements[id];
+    for (let i = 0; i < this._sorted.length; i++) {
+      if (this._sorted[i] === id) {
+        this._sorted.splice(i, 1);
+        break;
+      }
+    }
+    this.layout();
+    return element;
   }
 
   clear(): void {
@@ -274,7 +278,7 @@ export default class WidgetList extends HTMLElement {
 
     let x = this._xinc1 < 0 ? -this._elementwidth : 0;
     let y = this._yinc1 < 0 ? -this._elementheight : 0;
-    let rowcolindex = 0;
+    let rowColIndex = 0;
     let count = 0;
 
     this._sorted.forEach((id: number) => {
@@ -294,22 +298,21 @@ export default class WidgetList extends HTMLElement {
       }
       container.style.display = 'block';
 
-
       count++;
 
       container.style.position = 'absolute';
-      container.style.left = String(x);
-      container.style.top = String(y);
+      container.style.left = x.toString();
+      container.style.top = y.toString();
 
       x = x + (this._xinc1 * this._elementwidth);
       y = y + (this._yinc1 * this._elementheight);
-      rowcolindex = rowcolindex + 1;
-      if (rowcolindex === this._rowcolsize) {
-        x = x - (this._xinc1 * this._elementwidth) * rowcolindex;
-        y = y - (this._yinc1 * this._elementheight) * rowcolindex;
+      rowColIndex = rowColIndex + 1;
+      if (rowColIndex === this._rowcolsize) {
+        x = x - (this._xinc1 * this._elementwidth) * rowColIndex;
+        y = y - (this._yinc1 * this._elementheight) * rowColIndex;
         x = x + (this._xinc2 * this._elementwidth);
         y = y + (this._yinc2 * this._elementheight);
-        rowcolindex = 0;
+        rowColIndex = 0;
       }
     });
   }
```

Was the issue missing code in `removeElement`? And the rest of this is additional cleanup?

File: resources/widget_list.ts (hunk 1 above)
Comment (Trim21) [reply resources/widget_list.ts:24]: `this.shadowRoot` is `ShadowRoot|null` and requires additional null checker when using it.

**Comment by Trim21** [comment]
> Was the issue missing code in `removeElement`? And the rest of this is additional cleanup?

Yes, other diff are minor.

File: resources/widget_list.ts (hunk 1 above)
Comment (quisquous) [reply resources/widget_list.ts:24]: Ah, thanks.  I think there is only one other call for `this.shadow.`, so I think it'd be cleaner to turn that into `this.shadowRoot.?` and not have the extra member variable to think about here.

Comment (Trim21) [reply resources/widget_list.ts:24]: Yes, that's reasonable.

Comment (quisquous) [reply resources/widget_list.ts:24]: Would you mind doing that, and then we can merge this?

Comment (Trim21) [reply resources/widget_list.ts:24]: busy in new game version these days.

Comment (quisquous) [reply resources/widget_list.ts:24]: No worries! I just wanted to make sure you weren't waiting on me.  :)

**Pull request merged**